### PR TITLE
Configuration property to work with database schemas generated for Hibernate ORM 5.6

### LINKED ISSUE #31475: Configuration property to work with database schemas generated for Hibernate ORM 5.6
### Description

Hibernate ORM 6 changes a few defaults regarding database schemas, and that could be a problem for pre-existing applications that migrate to Quarkus 3 + ORM 6.

We need to add a unique, per-PU property that would switch several configuration properties to their "legacy" value so that Hibernate ORM 6 continues to work with those "legacy" database schemas.

Ideally, those configuration properties should be overrideable through `quarkus.hibernate-orm.unsupported-properties`, so that people can iterate when migrating.

The exact list of properties set by that will be automatically overridden needs to be documented somewhere; though I believe a link to the source would be enough, if the source is a single class that is focused only on this overriding and isn't overly complex.

### Implementation ideas

E.g.:

```
quarkus.hibernate-orm.schema.compatibility = 5.6
```

### Limitations

For now, we will have a global property for all persistence units but that seems like an acceptable limitation.

This compatibility property is specifically about the database schema. It doesn't improve the compatibility for annotations and API changes.

### List of compatibility issues

- [x] [Sequences](https://github.com/hibernate/hibernate-orm/blob/6.0/migration-guide.adoc#implicit-identifier-sequence-and-table-name): - we need to set it to `legacy` to use `hibernate_sequence` by default
- [x] [Duration mapping changes](https://github.com/hibernate/hibernate-orm/blob/6.0/migration-guide.adoc#duration-mapping-changes) 
- [x] [UUID mapping changes](https://github.com/hibernate/hibernate-orm/blob/6.0/migration-guide.adoc#uuid-mapping-changes):  **AND** [mariadb](https://github.com/hibernate/hibernate-orm/blob/6.2/migration-guide.adoc#uuid-mapping-changes-on-mariadb) **AND** [sql-server](https://github.com/hibernate/hibernate-orm/blob/6.2/migration-guide.adoc#uuid-mapping-changes-on-sql-server)
- [x] [Instant mapping changes](https://github.com/hibernate/hibernate-orm/blob/6.0/migration-guide.adoc#instant-mapping-changes)
- [ ] [Basic array/collection mapping](https://github.com/hibernate/hibernate-orm/blob/6.1/migration-guide.adoc#basic-arraycollection-mapping):  we are still undecided of if we want to do something here. It is not high priority: people can annotate their properties to work around the problem and it should not be very common
- [x] [Enum mapping changes](https://github.com/hibernate/hibernate-orm/blob/6.1/migration-guide.adoc#enum-mapping-changes ) - Christian said it should be backward compatible
- [x] [Datatype for enums](https://github.com/hibernate/hibernate-orm/blob/6.2/migration-guide.adoc#datatype-for-enums ): Christian said it should be backward compatible
- [x] [Timezone and offset storage](https://github.com/hibernate/hibernate-orm/blob/6.2/migration-guide.adoc#timezone-and-offset-storage):  Apparently not much to do here but might require some clarifications
- [ ] #31588

### PR DESCRIPTION
**Pull request by yrodiere**
Fixes #31475.

This is far from perfect but should addresses the biggest pain points.

I'll stress once again (hopefully it's clear in the docs): this is a *best-effort* feature. The goal is *not* to provide perfect compatibility for everything, but just to ease migration and early testing by allowing application developers to delay work on some of the most obvious incompatibilities when it comes to the database schema.

Note this only addresses what we could find in the migration guides; anything missing from migration guides hasn't been addressed.

The main problem remaining is enums mapped as ordinals (which is the default): their schema definition is different in Hibernate ORM 6 for pretty much every single dialect, resulting in schema validation. This will likely affect a large proportion of existing applications. Unfortunately we don't have a setting to revert to ORM 5 ("legacy") behavior for enums. Maybe we should add one in ORM 6.2?

---

The property can be used like this:

```
quarkus.hibernate-orm.database.orm-compatibility.version=5.6
```

If you disagree with the name, please provide constructive suggestions.

---

Regarding documentation, I'd personally stick to documenting the configuration property + referencing the property in the migration guide. I would not add a dedicated section in the Hibernate ORM Extension guide, because we want to remove this property eventually.

### CODE REVIEW

**Comment by Sanne** [comment]
Wow you did a lot of work already and looks very promising!

Let me start with high level questions, as I have to look better soon:

Do you have a phase-out plan for this property? I'm wondering how we'll handle when *some* of the subset flag you plan to apply will cease to work.

Also - should we nag users about its use? for example logging a warning?

Both aspects could be addressed by logging the individual flags you're setting as a warning;

```
WARN: Hibernate ORM is configured with `hibernate.id.db_structure_naming_strategy=legacy` because of `quarkus.hibernate-orm.database.orm-compatibility.version=5.6` - this affects schema compatibility, and the flag `hibernate.id.db_structure_naming_strategy` is not flagged as stable API.
WARN: Hibernate ORM is configured with `hibernate.type.preferred_instant_jdbc_type=TIMESTAMP` because of `quarkus.hibernate-orm.database.orm-compatibility.version=5.6` - this affects schema compatibility, and the flag `hibernate.type.preferred_instant_jdbc_type` is not flagged as stable API.
[...]
```
That would encourage people to get rid of them soonish, and also give them better awareness of what's going on behind the scenes; could be important if some of these incubating/deprecated propertie has to phase out or be adapted in some way.

**Comment by yrodiere** [comment]
> Wow you did a lot of work already and looks very promising!

It helped that @gsmet went through the ORM migration guides and listed the most problematic breaking changes :)

> Do you have a phase-out plan for this property?

In my mind we'd deprecate it soon (3.1? 3.2? Whatever makes sense for product needs.)

> I'm wondering how we'll handle when some of the subset flag you plan to apply will cease to work.

Surely that won't happen before the next major of Hibernate ORM. I trust we don't want to introduce breaking changes in minors :)

> Also - should we nag users about its use? for example logging a warning?
> 
> Both aspects could be addressed by logging the individual flags you're setting as a warning;
> 
> ```
> WARN: Hibernate ORM is configured with `hibernate.id.db_structure_naming_strategy=legacy` because of `quarkus.hibernate-orm.database.orm-compatibility.version=5.6` - this affects schema compatibility, and the flag `hibernate.id.db_structure_naming_strategy` is not flagged as stable API.
> WARN: Hibernate ORM is configured with `hibernate.type.preferred_instant_jdbc_type=TIMESTAMP` because of `quarkus.hibernate-orm.database.orm-compatibility.version=5.6` - this affects schema compatibility, and the flag `hibernate.type.preferred_instant_jdbc_type` is not flagged as stable API.
> [...]
> ```
> 
> That would encourage people to get rid of them soonish, and also give them better awareness of what's going on behind the scenes; could be important if some of these incubating/deprecated propertie has to phase out or be adapted in some way.

That's a very good idea. I'll add something along these lines.

**Comment by Sanne** [comment]
> Surely that won't happen before the next major of Hibernate ORM. I trust we don't want to introduce breaking changes in minors :)

I suspect you're right, but bear in mind some of these are flagged as "@Incubating" .. zero promises.

**Comment by yrodiere** [comment]
I made the tests resilient to different timezone changes (see last commit for an explanation), so that should fix CI.

@Sanne you can see the problems related to default sequence increment by commenting out the line that forces 5.6 compatibility in `integration-tests/hibernate-orm-compatibility-5.6/postgresql/src/main/resources/application.properties` and then running this:

```
./mvnw  -f integration-tests/hibernate-orm-compatibility-5.6/postgresql clean install -Dstart-containers -Dtest-containers
```

This will give you the error mentioned in the 4th commit's message:

```
[ERROR]   CompatibilityTest.sequence_genericGenerator_defaultAllocation:53
expected: 3L
 but was: -46L
```

**Comment by yrodiere** [comment]
@Sanne @gsmet IMO this is ready to merge, but if we want to merge this before Alpha5 I'll need an approval before wednesday (and preferably sooner if there are requests for changes).

So... WDYT? :)

**Comment by yrodiere** [comment]
Note I also started a [migration guide from ORM 5 to 6 in Quarkus](https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0:-Hibernate-ORM-5-to-6-migration) that I think addresses the main pain points (at least it addresses those @gsmet noticed). I suppose we can always dig deeper but I think it's a good start.

**Comment by Sanne** [comment]
I'll look now, sorry for the delay.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
Comment (Sanne) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:387]: This confuses me a bit - why the first one?

**Review by Sanne — APPROVED** [review]
Great work

**Comment by Sanne** [comment]
ah I see now you added the test in native-image form as well; do you see value in it?

Otherwise I'd avoid it, saving some energy

**Comment by Sanne** [comment]
> @Sanne you can see the problems related to default sequence increment by commenting out the line that forces 5.6 compatibility in `integration-tests/hibernate-orm-compatibility-5.6/postgresql/src/main/resources/application.properties` and then running this: [..]

Thanks yes that was useful. But did you notice that if you validate the schema Hibernate points out the problem:

```
Caused by: org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: sequence [gengendefallocsize] defined inconsistent increment-size; found [1] but expecting [50]
```

That's really nice, and will help mitigate some confusion as well.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
Comment (yrodiere) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:387]: No particular reason, `findAny` (if that exists) would work. I copy pasted that code from the method that handles non-persistence.xml config.

**Comment by yrodiere** [comment]
> Otherwise I'd avoid it, saving some energy

Ok, will do.

> But did you notice that if you validate the schema Hibernate points out the problem:

Right, that's nice. We should probably hunt down examples that use `generation = none` though, I think there's a few. And obviously that doesn't solve the problem of current values (which must be incremented to cope with the `pooled` optimizer's behavior). But at least there's a heads-up.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:387]: But don't we attempt to match the right datasource configuration for this PU ?

**Comment by Sanne** [comment]
BTW don't consider my comments at all blocking, feel free to merge and address the minor points another time. Leaving that up to you.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
Comment (yrodiere) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:387]: The code for `persistence.xml` handling has always assumed we were using the default datasource for these PUs:

https://github.com/yrodiere/quarkus/blob/8f171cdbcfa4c810f5ec10377240a5f3278607c0/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L396-L399

This new code is in line with that assumption.

If you think it's wrong and we should extract the datasource name from the `persistence.xml` (I guess it can be configured there?), I can open another ticket?

Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:387]: ok, existing code - no rush then. But yes it surprised me a bit - I guess I forgot we ignore the datasource section from a persistence.xml.
I guess ignoring it is fine, but to make things better we could try to spot if the user defined a datasource section and warn about it being ignored / or even fail.

**Pull request merged**
